### ISSUE TITLE
Changeset types should be generic

### ISSUE THREAD

**Issue by vlascik** [validated-changeset/validated-changeset#173]
It would help the usability of changesets in Typescript if types for changesets were generic, e.g. `BufferedChangeset<UserModel>`, that way we would have the proper type safety and wouldn't have to rely on type casting constantly.

**Comment by vlascik** [comment]
Right, I imagine typing proxies could be an issue (never tried it), but seems that people have come up with some workarounds (e.g. here https://stackoverflow.com/questions/59390026/how-to-use-es6-proxy-in-typescript). Can you maybe try to flesh it out, and if you have issues maybe we can summon the gods of typescript at #topic-typescript? :D

**Comment by bakerac4** [comment]
1. I was just coming here to add an issue like this, so Im glad Im not the only one.
2. @vlascik I use this in our validation addon built on ember-changeset/ember-changeset validations and it works fairly well. Obviously doesn't provide even close to property type safety but for most use cases thats all we need it for.

`export type GenericChangeset<T> = BufferedChangeset & { data: T }`

**Comment by miguelcobain** [comment]
@bakerac4 those types were a great help! Thank you.

For reference, this is what I ended up with in my `types/index.d.ts`:

```ts
// fix changeset types
// taken from https://github.com/validated-changeset/validated-changeset/issues/173
import type Model from '@ember-data/model';
import type { Changeset } from 'ember-changeset';

export type ModelAttributes<T extends Model> = Pick<T, Exclude<keyof T, keyof Model>>;
export type ModelProperties<T extends Model> = Readonly<Pick<T, Exclude<keyof T, keyof ModelAttributes<T>>>>;
export type ChangesetAttributes<T> = T extends Model ? ModelAttributes<T> & ModelProperties<T> : T;
export type BufferedChangeset = ReturnType<typeof Changeset>;
export type GenericChangeset<T> = BufferedChangeset & ChangesetAttributes<T> & { data: T };
```

and then using it in a component

```ts
import { Changeset } from 'ember-changeset';

import type MyModel from 'my-app/models/my-model';
import type { GenericChangeset } from 'index';

class EditModel extends Component<PropertyGeneralSignature> {
  changeset = Changeset(this.args.myModel) as GenericChangeset<MyModel>;
```

would be great if ember-changeset could have this return type instead.

What do you think @snewcomer ?
